hs_core/irods.py

```diff
@@ -0,0 +1,246 @@
+        """ Return the home path for local iRODS resources """
+        return settings.IRODS_CWD
+
+    def update_bag(self):
+        """
+        Update a bag if necessary.
+
+        This uses the Django signal pre_check_bag_flag to prepare collections, i
+        and then checks the AVUs 'metadata_dirty' and 'bag_modified' to determine
+        whether to regenerate the metadata files and/or bag.
+
+        This is a synchronous update. The call waits until the update is finished.
+        """
+        from hs_core.tasks import create_bag_by_irods
+        from hs_core.hydroshare.resource import check_resource_type
+        from hs_core.hydroshare.hs_bagit import create_bag_files
+
+        # send signal for pre_check_bag_flag
+        resource_cls = check_resource_type(self.resource_type)
+        pre_check_bag_flag.send(sender=resource_cls, resource=self)
+
+        metadata_dirty = self.getAVU('metadata_dirty')
+        bag_modified = self.getAVU('bag_modified')
+
+        if metadata_dirty:  # automatically cast to Bool
+            create_bag_files(self)
+            self.setAVU('metadata_dirty', False)
+
+        # the ticket system does synchronous bag creation.
+        # async bag creation isn't supported.
+        if bag_modified:  # automatically cast to Bool
+            create_bag_by_irods(self.short_id)
+            self.setAVU('bag_modified', False)
+
+    def update_metadata_files(self):
+        """
+        Make the metadata files resourcemetadata.xml and resourcemap.xml up to date.
+
+        This checks the "metadata dirty" AVU before updating files if necessary.
+        """
+        from hs_core.hydroshare.hs_bagit import create_bag_files
+
+        metadata_dirty = self.getAVU('metadata_dirty')
+        if metadata_dirty:
+            create_bag_files(self)
+            self.setAVU('metadata_dirty', False)
+
+    def create_ticket(self, user, path=None, write=False, allowed_uses=1):
+        """
+        create an iRODS ticket for reading or modifying a resource
+
+        :param user: user to authorize
+        :param path: path in iRODS to the object being requested.
+        :return:
+
+        :raises PermissionDenied: if user is not allowed to create the ticket.
+        :raises SessionException: if ticket fails to be created for some reason.
+        :raises SuspiciousFileOperation: if path uses .. illegally
+
+        WARNING: This creates a ticket that expires in one hour in UTC. If the
+        iRODS and django servers are in different time zones and not set for UTC,
+        this results in a useless ticket. This includes federated servers.
+
+        THERE IS NO STRAIGHTFORWARD MECHANISM IN IRODS for determining the time zone
+        or local time of an iRODS server.
+
+        Also, note that there is no mechanism for asynchronous bag creation in the
+        ticketing system. The bag is created *synchronously* if required. The
+        ticket is not issued until the bag exists.
```

Comment:
I might add that a more usable protocol for asynchronous bag creation is a separate issue. Our current protocol doesn't handle asynchrony well. Since I am proving a concept, I have avoided that issue by using the synchronous option. I need to fix that, but it is not simple to fix, and @Castronova  is waiting for my solution to **this** problem.